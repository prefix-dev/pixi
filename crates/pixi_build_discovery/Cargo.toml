[package]
authors.workspace = true
description = "Discover the build backend from a source tree"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "pixi_build_discovery"
readme.workspace = true
repository.workspace = true
version = "0.1.0"

[dependencies]
dunce = { workspace = true }
itertools = { workspace = true }
miette = { workspace = true }
pathdiff = { workspace = true }
serde = { workspace = true, optional = true, features = ["derive"] }
serde_json = { workspace = true }
thiserror = { workspace = true }
url = { workspace = true }

rattler_conda_types = { workspace = true }

pixi_build_type_conversions = { workspace = true }
pixi_build_types = { workspace = true }
pixi_manifest = { workspace = true }
pixi_spec = { workspace = true }
pixi_spec_containers = { workspace = true }

[dev-dependencies]
insta = { workspace = true, features = [
  "yaml",
  "filters",
  "glob",
  "redactions",
] }
pixi_build_discovery = { path = ".", features = ["serde"] }
pixi_test_utils = { workspace = true }
