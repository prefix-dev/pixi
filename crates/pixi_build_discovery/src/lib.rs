//! A crate that is responsible for discovering the build backend that should
//! be used for a particular source tree.
//!
//! # Example
//!
//! ```rust,no_run
//! use pixi_build_discovery::{DiscoveredBackend, EnabledProtocols};
//! use rattler_conda_types::ChannelConfig;
//! use std::path::{Path, PathBuf};
//!
//! // Set up the channel configuration (typically from your environment)
//! let channel_config = ChannelConfig::default_with_root_dir(PathBuf::from("."));
//!
//! // Enable all protocols (default)
//! let enabled_protocols = EnabledProtocols::default();
//!
//! // Path to the source directory or manifest
//! let source_path = Path::new("/path/to/project");
//!
//! // Attempt to discover the backend
//! match DiscoveredBackend::discover(source_path, &channel_config, &enabled_protocols) {
//!     Ok(backend) => println!("Discovered backend: {:?}", backend.backend_spec),
//!     Err(e) => eprintln!("Failed to discover backend: {e}"),
//! }
//! ```
//!
//! This crate provides types and logic to identify and describe build backends for projects
//! using pixi, or rattler-build based workflows. It supports discovery from both `pixi.toml`
//! and `recipe.yaml` manifests, and can be extended to support additional protocols.

mod backend_spec;
mod discovery;

pub use backend_spec::{
    BackendSpec, CommandSpec, EnvironmentSpec, JsonRpcBackendSpec, SystemCommandSpec,
};
pub use discovery::{
    BackendInitializationParams, DiscoveredBackend, DiscoveryError, EnabledProtocols,
};
