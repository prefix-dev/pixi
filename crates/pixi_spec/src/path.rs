use std::path::Path;

use rattler_conda_types::{package::ArchiveIdentifier, NamelessMatchSpec};
use typed_path::{Utf8NativePathBuf, Utf8TypedPathBuf};

use crate::SpecConversionError;

/// A specification of a package from a git repository.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct PathSpec {
    /// The path to the package
    pub path: Utf8TypedPathBuf,
}

impl PathSpec {
    /// Converts this instance into a [`NamelessMatchSpec`] if the path points
    /// to binary archive.
    pub fn try_into_nameless_match_spec(
        self,
        root_dir: &Path,
    ) -> Result<Option<NamelessMatchSpec>, SpecConversionError> {
        if !self.is_binary() {
            // Not a binary package
            return Ok(None);
        }

        // Convert the path to an absolute path based on the root_dir
        let path = if self.path.is_absolute() {
            self.path
        } else if let Ok(user_path) = self.path.strip_prefix("~/") {
            let home_dir = dirs::home_dir()
                .ok_or_else(|| SpecConversionError::InvalidPath(self.path.to_string()))?;
            let Some(home_dir_str) = home_dir.to_str() else {
                return Err(SpecConversionError::NotUtf8RootDir(home_dir));
            };
            Utf8TypedPathBuf::from(home_dir_str)
                .join(user_path)
                .normalize()
        } else {
            let Some(root_dir_str) = root_dir.to_str() else {
                return Err(SpecConversionError::NotUtf8RootDir(root_dir.to_path_buf()));
            };
            let native_root_dir = Utf8NativePathBuf::from(root_dir_str);
            if !native_root_dir.is_absolute() {
                return Err(SpecConversionError::NonAbsoluteRootDir(
                    root_dir.to_path_buf(),
                ));
            }

            native_root_dir.to_typed_path().join(self.path).normalize()
        };

        // Convert the absolute url to a file:// url
        let local_file_url =
            file_url::file_path_to_url(path.to_path()).expect("failed to convert path to file url");

        Ok(Some(NamelessMatchSpec {
            url: Some(local_file_url),
            ..NamelessMatchSpec::default()
        }))
    }

    /// Converts this instance into a [`PathSourceSpec`] if the path points to a
    /// source package. Otherwise, returns this instance unmodified.
    #[allow(clippy::result_large_err)]
    pub fn try_into_source_path(self) -> Result<PathSourceSpec, Self> {
        if self.is_binary() {
            Err(self)
        } else {
            Ok(PathSourceSpec { path: self.path })
        }
    }

    /// Returns true if this path points to a binary archive.
    pub fn is_binary(&self) -> bool {
        self.path
            .file_name()
            .and_then(ArchiveIdentifier::try_from_path)
            .is_some()
    }
}

/// Path to a source package. Different from [`PathSpec`] in that this type only
/// refers to source packages.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct PathSourceSpec {
    /// The path to the package. Either a directory or an archive.
    pub path: Utf8TypedPathBuf,
}

impl From<PathSourceSpec> for PathSpec {
    fn from(value: PathSourceSpec) -> Self {
        Self { path: value.path }
    }
}
