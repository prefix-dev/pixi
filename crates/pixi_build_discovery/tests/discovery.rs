use insta::{assert_snapshot, assert_yaml_snapshot, glob};
use pixi_build_discovery::{DiscoveredBackend, EnabledProtocols};
use rattler_conda_types::ChannelConfig;
use std::path::{Path, PathBuf};

fn discovery_directory() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR")).join("../../tests/data/discovery")
}

/// This macro is used to assert the discovery of a backend and compare it with a snapshot.
///
/// Errors are also handled and compared with a snapshot.
macro_rules! assert_discover_snapshot {
    ($path:expr) => {
        let file_path = Path::new(file!()).parent().unwrap();
        let channel_config = ChannelConfig::default_with_root_dir(file_path.to_owned());

        // Run the discovery on the input path
        match DiscoveredBackend::discover($path, &channel_config, &EnabledProtocols::default()) {
            Ok(backend) => {
                assert_yaml_snapshot!(backend,
                // Perform some redaction on fields that contain paths. We need to make them cross-platform compatible.
                {
                    "[\"init-params\"][\"manifest-path\"]" => insta::dynamic_redaction(|value, _path| {
                        value.as_str().unwrap().replace("\\", "/")
                     }),
                    "[\"init-params\"][\"source-dir\"]" => "[SOURCE_PATH]",
                });
            }
            Err(err) => {
                assert_snapshot!(pixi_test_utils::format_diagnostic(&err));
            }
        }
    };
}

/// A test to check what discovery looks like for different use cases.
///
/// The test cases are located in the `tests/data/discovery` directory. Every
/// directory (or subdirectory) that contains a file TEST-CASE is used as a
/// test case.
#[test]
fn test_discovery() {
    glob!("../../../tests/data/discovery", "**/TEST-CASE", |path| {
        let path = dunce::canonicalize(path.parent().unwrap()).unwrap();
        let new_suffix = insta::Settings::clone_current()
            .snapshot_suffix()
            .unwrap()
            .strip_suffix("TEST-CASE")
            .unwrap()
            .trim_end_matches(['/', '\\'])
            .to_owned();
        let source_path_regex = path.to_string_lossy().replace(r"\", r"\\");
        insta::with_settings!({
            filters => vec![
                (source_path_regex.as_str(), "file://<ROOT>"),
                (r"\\", r"/"),
            ],
            snapshot_suffix => new_suffix}, {
            assert_discover_snapshot!(&path);
        })
    });
}

#[test]
fn test_non_existing() {
    assert_discover_snapshot!(Path::new("/non-existing"));
}

#[test]
fn test_direct_recipe() {
    let path = dunce::canonicalize(discovery_directory().join("recipe_yaml/recipe.yaml")).unwrap();
    let source_path_regex = path
        .parent()
        .unwrap()
        .to_string_lossy()
        .replace(r"\", r"\\\\");
    insta::with_settings!({
        filters => vec![
            (source_path_regex.as_str(), "file://<ROOT>"),
            (r"\\", r"/"),
        ],
    }, {
        assert_discover_snapshot ! ( & path);
    });
}
