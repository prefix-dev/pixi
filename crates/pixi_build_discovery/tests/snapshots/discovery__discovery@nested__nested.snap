---
source: crates/pixi_build_discovery/tests/discovery.rs
expression: backend
input_file: tests/data/discovery/nested/nested/TEST-CASE
---
backend-spec:
  type: json-rpc
  name: pixi_build_backend
  command:
    type: environment-spec
    requirement:
      - pixi_build_backend
      - version: "*"
    channels:
      - "https://prefix.dev/pixi-build-backends"
init-params:
  workspace-root: "file://<ROOT>/nested"
  source: ~
  source-anchor: "file://<ROOT>/nested/nested"
  manifest-path: "file://<ROOT>/nested/nested/pixi.toml"
  project-model:
    name: simple
    version: 0.1.0
    description: ~
    authors: ~
    license: ~
    licenseFile: ~
    readme: ~
    homepage: ~
    repository: ~
    documentation: ~
    targets:
      defaultTarget:
        hostDependencies: {}
        buildDependencies: {}
        runDependencies: {}
      targets: {}
  configuration: ~
  target-configuration: ~
